// oneai/lib/src/index.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var pipeline_1 = __importDefault(require("./pipeline"));
var clustering_1 = require("./clustering");
var skills_1 = require("./skills");
var conversation_1 = require("./parsing/conversation");
var srt_1 = __importDefault(require("./parsing/srt"));
var pipeline_2 = __importDefault(require("./api/pipeline"));
var logging_1 = __importDefault(require("./logging"));
var clustering_2 = __importDefault(require("./api/clustering"));
var OneAI = /** @class */ (function () {
    function OneAI(apiKey, params) {
        /**
         * A Language Skill is a package of trained NLP models.
         * Skills accept text and respond with processed texts and extracted metadata.
         *
         * Process texts with Skills using `Pipeline`s
         */
        this.skills = skills_1.skills;
        this.parsing = {
            parseConversation: conversation_1.parseConversation,
            toSRT: srt_1.default,
        };
        this.params = __assign(__assign(__assign({}, OneAI.defaultParams), params !== undefined && params), apiKey !== undefined && { apiKey: apiKey });
        this.logger = new logging_1.default();
        this.pipelineApiClient = new pipeline_2.default(this.params, this.logger);
        this.Pipeline = (0, pipeline_1.default)(this.pipelineApiClient);
        this.clusteringApiClient = new clustering_2.default(this.params, this.logger);
        var Collection = (0, clustering_1.createCollectionClass)(this.clusteringApiClient);
        this.clustering = {
            Item: clustering_1.Item,
            Phrase: clustering_1.Phrase,
            Cluster: clustering_1.Cluster,
            Collection: Collection,
            getCollections: Collection.getCollections,
        };
    }
    OneAI.prototype.toString = function () {
        var _a;
        var apiKeyString = (((_a = this.params.apiKey) === null || _a === void 0 ? void 0 : _a.length) >= 8) ? this.params.apiKey.substring(0, 8) + '*'.repeat(this.params.apiKey.length - 8) : undefined;
        return "One AI Client - API Key ".concat(apiKeyString);
    };
    OneAI.skills = skills_1.skills;
    OneAI.defaultParams = {
        apiKey: '',
        timeout: 60,
        baseURL: 'https://api.oneai.com',
    };
    OneAI.OneAI = OneAI;
    return OneAI;
}());
module.exports = OneAI;
